File: packages/Axiom.jl/zig/src/norm.zig
// SPDX-License-Identifier: PMPL-1.0-or-later
//! Normalization Operations
//!
//! Layer normalization, RMS normalization, and batch normalization.

const std = @import("std");
const math = std.math;

const VEC_SIZE = 8;
const Vec = @Vector(VEC_SIZE, f32);

/// Layer Normalization
/// Normalizes over the last dimension (hidden_size)
/// SIMD-optimized mean, variance, and normalization.
pub fn layernorm(
    x_ptr: [*]const f32,
    y_ptr: [*]f32,
    gamma_ptr: [*]const f32,
    beta_ptr: [*]const f32,
    batch_size: usize,
    hidden_size: usize,
    eps: f32,
) void {
    const hidden_f: f32 = @floatFromInt(hidden_size);

    var b: usize = 0;
    while (b < batch_size) : (b += 1) {
        const x = x_ptr[b * hidden_size ..][0..hidden_size];
        var y = y_ptr[b * hidden_size ..][0..hidden_size];

        // Compute mean with SIMD
        var sum_acc: Vec = @splat(0);
        var i: usize = 0;
        while (i + VEC_SIZE <= hidden_size) : (i += VEC_SIZE) {
            const x_vec: Vec = x[i..][0..VEC_SIZE].*;
            sum_acc += x_vec;
        }
        var sum: f32 = @reduce(.Add, sum_acc);
        while (i < hidden_size) : (i += 1) {
            sum += x[i];
        }
        const mean = sum / hidden_f;

        // Compute variance with SIMD
        const mean_vec: Vec = @splat(mean);
        var var_acc: Vec = @splat(0);
        i = 0;
        while (i + VEC_SIZE <= hidden_size) : (i += VEC_SIZE) {
            const x_vec: Vec = x[i..][0..VEC_SIZE].*;
            const diff = x_vec - mean_vec;
            var_acc += diff * diff;
        }
        var var_sum: f32 = @reduce(.Add, var_acc);
        while (i < hidden_size) : (i += 1) {
            const diff = x[i] - mean;
            var_sum += diff * diff;
        }
        const inv_std = 1.0 / @sqrt(var_sum / hidden_f + eps);

        // Normalize and scale with SIMD
        const inv_std_vec: Vec = @splat(inv_std);
        i = 0;
        while (i + VEC_SIZE <= hidden_size) : (i += VEC_SIZE) {
            const x_vec: Vec = x[i..][0..VEC_SIZE].*;
            const g_vec: Vec = gamma_ptr[i..][0..VEC_SIZE].*;
            const b_vec: Vec = beta_ptr[i..][0..VEC_SIZE].*;
            const normalized = (x_vec - mean_vec) * inv_std_vec;
            y[i..][0..VEC_SIZE].* = g_vec * normalized + b_vec;
        }
        while (i < hidden_size) : (i += 1) {
            const normalized = (x[i] - mean) * inv_std;
            y[i] = gamma_ptr[i] * normalized + beta_ptr[i];
        }
    }
}

/// RMS Normalization (used in LLaMA, etc.)
/// x_norm = x / sqrt(mean(x^2) + eps) * weight
pub fn rmsnorm(
    x_ptr: [*]const f32,
    y_ptr: [*]f32,
    weight_ptr: [*]const f32,
    batch_size: usize,
    hidden_size: usize,
    eps: f32,
) void {
    const hidden_f: f32 = @floatFromInt(hidden_size);

    var b: usize = 0;
    while (b < batch_size) : (b += 1) {
        const x = x_ptr[b * hidden_size ..][0..hidden_size];
        var y = y_ptr[b * hidden_size ..][0..hidden_size];

        // Compute mean of squares with SIMD
        var sum_sq: f32 = 0;
        var i: usize = 0;

        var acc: Vec = @splat(0);
        while (i + VEC_SIZE <= hidden_size) : (i += VEC_SIZE) {
            const x_vec: Vec = x[i..][0..VEC_SIZE].*;
            acc += x_vec * x_vec;
        }
        sum_sq = @reduce(.Add, acc);

        // Remainder
        while (i < hidden_size) : (i += 1) {
            sum_sq += x[i] * x[i];
        }

        const rms = @sqrt(sum_sq / hidden_f + eps);
        const inv_rms = 1.0 / rms;

        // Scale with SIMD
        const inv_rms_vec: Vec = @splat(inv_rms);
        i = 0;
        while (i + VEC_SIZE <= hidden_size) : (i += VEC_SIZE) {
            const x_vec: Vec = x[i..][0..VEC_SIZE].*;
            const w_vec: Vec = weight_ptr[i..][0..VEC_SIZE].*;
            y[i..][0..VEC_SIZE].* = x_vec * inv_rms_vec * w_vec;
        }

        // Remainder
        while (i < hidden_size) : (i += 1) {
            y[i] = x[i] * inv_rms * weight_ptr[i];
        }
    }
}

/// Batch Normalization
/// For inference mode (uses running statistics)
pub fn batchnorm(
    x_ptr: [*]const f32,
    y_ptr: [*]f32,
    gamma_ptr: [*]const f32,
    beta_ptr: [*]const f32,
    running_mean_ptr: [*]const f32,
    running_var_ptr: [*]const f32,
    batch_size: usize,
    num_features: usize,
    eps: f32,
) void {
    // Precompute inv_std for each feature
    var inv_std: [4096]f32 = undefined;
    var i: usize = 0;
    while (i < num_features) : (i += 1) {
        inv_std[i] = 1.0 / @sqrt(running_var_ptr[i] + eps);
    }

    var b: usize = 0;
    while (b < batch_size) : (b += 1) {
        const x = x_ptr[b * num_features ..][0..num_features];
        var y = y_ptr[b * num_features ..][0..num_features];

        var f: usize = 0;
        while (f < num_features) : (f += 1) {
            const normalized = (x[f] - running_mean_ptr[f]) * inv_std[f];
            y[f] = gamma_ptr[f] * normalized + beta_ptr[f];
        }
    }
}

/// Instance Normalization
pub fn instancenorm(
    x_ptr: [*]const f32,
    y_ptr: [*]f32,
    gamma_ptr: ?[*]const f32,
    beta_ptr: ?[*]const f32,
    batch: usize,
    h: usize,
    w: usize,
    channels: usize,
    eps: f32,
) void {
    const spatial_size = h * w;
    const spatial_f: f32 = @floatFromInt(spatial_size);

    var n: usize = 0;
    while (n < batch) : (n += 1) {
        var c: usize = 0;
        while (c < channels) : (c += 1) {
            // Compute mean over spatial dimensions
            var sum: f32 = 0;
            var s: usize = 0;
            while (s < spatial_size) : (s += 1) {
                const idx = n * spatial_size * channels + s * channels + c;
                sum += x_ptr[idx];
            }
            const mean = sum / spatial_f;

            // Compute variance
            var var_sum: f32 = 0;
            s = 0;
            while (s < spatial_size) : (s += 1) {
                const idx = n * spatial_size * channels + s * channels + c;
                const diff = x_ptr[idx] - mean;
                var_sum += diff * diff;
            }
            const variance = var_sum / spatial_f;
            const inv_std = 1.0 / @sqrt(variance + eps);

            // Get affine parameters
            const gamma = if (gamma_ptr) |g| g[c] else 1.0;
            const beta = if (beta_ptr) |b| b[c] else 0.0;

            // Normalize and scale
            s = 0;
            while (s < spatial_size) : (s += 1) {
                const idx = n * spatial_size * channels + s * channels + c;
                const normalized = (x_ptr[idx] - mean) * inv_std;
                y_ptr[idx] = gamma * normalized + beta;
            }
        }
    }
}

/// Group Normalization
pub fn groupnorm(
    x_ptr: [*]const f32,
    y_ptr: [*]f32,
    gamma_ptr: [*]const f32,
    beta_ptr: [*]const f32,
    batch: usize,
    h: usize,
    w: usize,
    channels: usize,
    num_groups: usize,
    eps: f32,
) void {
    const spatial_size = h * w;
    const channels_per_group = channels / num_groups;
    const group_size = spatial_size * channels_per_group;
    const group_size_f: f32 = @floatFromInt(group_size);

    var n: usize = 0;
    while (n < batch) : (n += 1) {
        var g: usize = 0;
        while (g < num_groups) : (g += 1) {
            const c_start = g * channels_per_group;

            // Compute mean over group
            var sum: f32 = 0;
            var s: usize = 0;
            while (s < spatial_size) : (s += 1) {
                var c = c_start;
                while (c < c_start + channels_per_group) : (c += 1) {
                    const idx = n * spatial_size * channels + s * channels + c;
                    sum += x_ptr[idx];
                }
            }
            const mean = sum / group_size_f;

            // Compute variance
            var var_sum: f32 = 0;
            s = 0;
            while (s < spatial_size) : (s += 1) {
                var c = c_start;
                while (c < c_start + channels_per_group) : (c += 1) {
                    const idx = n * spatial_size * channels + s * channels + c;
                    const diff = x_ptr[idx] - mean;
                    var_sum += diff * diff;
                }
            }
            const variance = var_sum / group_size_f;
            const inv_std = 1.0 / @sqrt(variance + eps);

            // Normalize and scale
            s = 0;
            while (s < spatial_size) : (s += 1) {
                var c = c_start;
                while (c < c_start + channels_per_group) : (c += 1) {
                    const idx = n * spatial_size * channels + s * channels + c;
                    const normalized = (x_ptr[idx] - mean) * inv_std;
                    y_ptr[idx] = gamma_ptr[c] * normalized + beta_ptr[c];
                }
            }
        }
    }
}

File: packages/Axiom.jl/rust/src/ops/norm.rs
//! Normalization operations

use ndarray::ArrayD;

/// Batch Normalization
///
/// x_norm = (x - mean) / sqrt(var + eps)
/// output = gamma * x_norm + beta
#[allow(clippy::too_many_arguments)]
pub fn batchnorm(
    input: &ArrayD<f32>,
    gamma: &[f32],
    beta: &[f32],
    running_mean: &mut [f32],
    running_var: &mut [f32],
    eps: f32,
    momentum: f32,
    training: bool,
) -> ArrayD<f32> {
    let shape = input.shape();
    let ndim = shape.len();
    let num_features = shape[ndim - 1];

    assert_eq!(gamma.len(), num_features);
    assert_eq!(beta.len(), num_features);

    let mut output = input.clone();

    if training {
        // Compute batch statistics
        let batch_size: usize = shape[..ndim - 1].iter().product();

        for c in 0..num_features {
            // Compute mean
            let mut sum = 0.0f32;
            for i in 0..batch_size {
                let idx = i * num_features + c;
                sum += input.as_slice().expect("norm: array not contiguous")[idx];
            }
            let mean = sum / batch_size as f32;

            // Compute variance
            let mut var_sum = 0.0f32;
            for i in 0..batch_size {
                let idx = i * num_features + c;
                let diff = input.as_slice().expect("norm: array not contiguous")[idx] - mean;
                var_sum += diff * diff;
            }
            let var = var_sum / batch_size as f32;

            // Update running statistics
            running_mean[c] = (1.0 - momentum) * running_mean[c] + momentum * mean;
            running_var[c] = (1.0 - momentum) * running_var[c] + momentum * var;

            // Normalize and scale
            let inv_std = 1.0 / (var + eps).sqrt();
            let output_slice = output.as_slice_mut().expect("norm: array not contiguous");
            for i in 0..batch_size {
                let idx = i * num_features + c;
                let normalized = (output_slice[idx] - mean) * inv_std;
                output_slice[idx] = gamma[c] * normalized + beta[c];
            }
        }
    } else {
        // Use running statistics
        let batch_size: usize = shape[..ndim - 1].iter().product();

        for c in 0..num_features {
            let inv_std = 1.0 / (running_var[c] + eps).sqrt();
            let output_slice = output.as_slice_mut().expect("norm: array not contiguous");

            for i in 0..batch_size {
                let idx = i * num_features + c;
                let normalized = (output_slice[idx] - running_mean[c]) * inv_std;
                output_slice[idx] = gamma[c] * normalized + beta[c];
            }
        }
    }

    output
}

/// Layer Normalization
///
/// Normalizes over the last N dimensions
pub fn layernorm(
    input: &ArrayD<f32>,
    gamma: &ArrayD<f32>,
    beta: &ArrayD<f32>,
    normalized_shape: &[usize],
    eps: f32,
) -> ArrayD<f32> {
    let shape = input.shape();
    let ndim = shape.len();
    let norm_ndim = normalized_shape.len();

    // Number of elements to normalize over
    let norm_size: usize = normalized_shape.iter().product();
    let batch_size: usize = shape[..ndim - norm_ndim].iter().product();

    let mut output = input.clone();

    let output_slice = output.as_slice_mut().expect("norm: array not contiguous");
    let input_slice = input.as_slice().expect("norm: array not contiguous");
    let gamma_slice = gamma.as_slice().expect("norm: array not contiguous");
    let beta_slice = beta.as_slice().expect("norm: array not contiguous");

    for batch in 0..batch_size {
        let start = batch * norm_size;
        let end = start + norm_size;

        // Compute mean
        let sum: f32 = input_slice[start..end].iter().sum();
        let mean = sum / norm_size as f32;

        // Compute variance
        let var: f32 = input_slice[start..end]
            .iter()
            .map(|&x| (x - mean).powi(2))
            .sum::<f32>()
            / norm_size as f32;

        // Normalize and scale
        let inv_std = 1.0 / (var + eps).sqrt();

        for i in 0..norm_size {
            let idx = start + i;
            let normalized = (output_slice[idx] - mean) * inv_std;
            output_slice[idx] = gamma_slice[i] * normalized + beta_slice[i];
        }
    }

    output
}

/// Instance Normalization
///
/// Normalizes each sample independently over spatial dimensions
pub fn instancenorm(
    input: &ArrayD<f32>, // (N, H, W, C) or (N, ..., C)
    gamma: Option<&[f32]>,
    beta: Option<&[f32]>,
    eps: f32,
) -> ArrayD<f32> {
    let shape = input.shape();
    let ndim = shape.len();
    let batch_size = shape[0];
    let num_channels = shape[ndim - 1];
    let spatial_size: usize = shape[1..ndim - 1].iter().product();

    let mut output = input.clone();

    let output_slice = output.as_slice_mut().expect("norm: array not contiguous");
    let input_slice = input.as_slice().expect("norm: array not contiguous");

    for n in 0..batch_size {
        for c in 0..num_channels {
            // Compute mean over spatial dimensions
            let mut sum = 0.0f32;
            for s in 0..spatial_size {
                let idx = n * spatial_size * num_channels + s * num_channels + c;
                sum += input_slice[idx];
            }
            let mean = sum / spatial_size as f32;

            // Compute variance
            let mut var_sum = 0.0f32;
            for s in 0..spatial_size {
                let idx = n * spatial_size * num_channels + s * num_channels + c;
                let diff = input_slice[idx] - mean;
                var_sum += diff * diff;
            }
            let var = var_sum / spatial_size as f32;

            // Normalize and scale
            let inv_std = 1.0 / (var + eps).sqrt();
            let g = gamma.map_or(1.0, |g| g[c]);
            let b = beta.map_or(0.0, |b| b[c]);

            for s in 0..spatial_size {
                let idx = n * spatial_size * num_channels + s * num_channels + c;
                let normalized = (output_slice[idx] - mean) * inv_std;
                output_slice[idx] = g * normalized + b;
            }
        }
    }

    output
}

/// Group Normalization
///
/// Divides channels into groups and normalizes within each group
pub fn groupnorm(
    input: &ArrayD<f32>, // (N, H, W, C)
    num_groups: usize,
    gamma: &[f32],
    beta: &[f32],
    eps: f32,
) -> ArrayD<f32> {
    let shape = input.shape();
    let ndim = shape.len();
    let batch_size = shape[0];
    let num_channels = shape[ndim - 1];
    let spatial_size: usize = shape[1..ndim - 1].iter().product();

    assert_eq!(
        num_channels % num_groups,
        0,
        "num_channels must be divisible by num_groups"
    );
    let channels_per_group = num_channels / num_groups;

    let mut output = input.clone();

    let output_slice = output.as_slice_mut().expect("norm: array not contiguous");
    let input_slice = input.as_slice().expect("norm: array not contiguous");

    for n in 0..batch_size {
        for g in 0..num_groups {
            let c_start = g * channels_per_group;
            let c_end = c_start + channels_per_group;

            // Compute mean over spatial and channel dimensions within group
            let mut sum = 0.0f32;
            let count = spatial_size * channels_per_group;

            for s in 0..spatial_size {
                for c in c_start..c_end {
                    let idx = n * spatial_size * num_channels + s * num_channels + c;
                    sum += input_slice[idx];
                }
            }
            let mean = sum / count as f32;

            // Compute variance
            let mut var_sum = 0.0f32;
            for s in 0..spatial_size {
                for c in c_start..c_end {
                    let idx = n * spatial_size * num_channels + s * num_channels + c;
                    let diff = input_slice[idx] - mean;
                    var_sum += diff * diff;
                }
            }
            let var = var_sum / count as f32;

            // Normalize and scale
            let inv_std = 1.0 / (var + eps).sqrt();

            for s in 0..spatial_size {
                for c in c_start..c_end {
                    let idx = n * spatial_size * num_channels + s * num_channels + c;
                    let normalized = (output_slice[idx] - mean) * inv_std;
                    output_slice[idx] = gamma[c] * normalized + beta[c];
                }
            }
        }
    }

    output
}

/// RMS Normalization (used in LLaMA, etc.)
///
/// x_norm = x / sqrt(mean(x^2) + eps)
/// output = weight * x_norm
pub fn rmsnorm(input: &ArrayD<f32>, weight: &[f32], eps: f32) -> ArrayD<f32> {
    let shape = input.shape();
    let ndim = shape.len();
    let hidden_size = shape[ndim - 1];
    let batch_size: usize = shape[..ndim - 1].iter().product();

    let mut output = input.clone();

    let output_slice = output.as_slice_mut().expect("norm: array not contiguous");
    let input_slice = input.as_slice().expect("norm: array not contiguous");

    for batch in 0..batch_size {
        let start = batch * hidden_size;
        let end = start + hidden_size;

        // Compute RMS
        let rms: f32 =
            input_slice[start..end].iter().map(|&x| x * x).sum::<f32>() / hidden_size as f32;
        let inv_rms = 1.0 / (rms + eps).sqrt();

        // Normalize and scale
        for (i, w) in weight.iter().enumerate().take(hidden_size) {
            let idx = start + i;
            output_slice[idx] = w * input_slice[idx] * inv_rms;
        }
    }

    output
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_layernorm() {
        let input = ArrayD::from_shape_vec(
            ndarray::IxDyn(&[2, 4]),
            vec![1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0],
        )
        .expect("test: shape mismatch");

        let gamma = ArrayD::from_elem(ndarray::IxDyn(&[4]), 1.0f32);
        let beta = ArrayD::from_elem(ndarray::IxDyn(&[4]), 0.0f32);

        let output = layernorm(&input, &gamma, &beta, &[4], 1e-5);

        // After layer norm, mean should be ~0, std should be ~1
        let first_row: Vec<f32> = output.slice(ndarray::s![0, ..]).iter().cloned().collect();
        let mean: f32 = first_row.iter().sum::<f32>() / 4.0;
        assert!((mean - 0.0).abs() < 0.01);
    }

    #[test]
    fn test_rmsnorm() {
        let input = ArrayD::from_shape_vec(
            ndarray::IxDyn(&[2, 4]),
            vec![1.0, 2.0, 3.0, 4.0, 1.0, 2.0, 3.0, 4.0],
        )
        .expect("test: shape mismatch");

        let weight = vec![1.0f32; 4];

        let output = rmsnorm(&input, &weight, 1e-5);

        // Output should be normalized
        assert!(output[[0, 0]].abs() < 1.0);
    }
}
